### PR TITLE
fix: master checks db newness before migrating [DET-10312]

### PR DESCRIPTION
**Pull request by jesse-amano-hpe**
## Ticket

[DET-10312]

## Description

Since creating the admin user without a password is a migration step, we need to check for invalid configuration (missing initial password) before the migration is run, but wait to apply the configuration until after the migration is run. This therefore splits the previous admin password configuration into two separate steps.

Due to the way this straddles the usual place Bun would be initialized, I didn't Bun-ify anything as part of this, and did a little bit of redundant config parsing to avoid doing anything that could call `initTheOneBun()` twice.

## Test Plan

This attempts to solve an issue that was found testing #9314 .

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-10312]: https://hpe-aiatscale.atlassian.net/browse/DET-10312?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/db/setup.go [hunk 1]
```diff
@@ -83,24 +85,64 @@ func Connect(opts *config.DBConfig) (*PgDB, error) {
...
+// incomplete state. If an error is returned, the bool should be ignored.
+func IsNew(opts *config.DBConfig) (bool, error) {
+	dbURL := fmt.Sprintf(cnxTpl, opts.User, opts.Password, opts.Host, opts.Port, opts.Name)
+	dbURL += fmt.Sprintf(sslTpl, opts.SSLMode, opts.SSLRootCert)
+	pgOpts, err := makeGoPgOpts(dbURL)
+	if err != nil {
+		return false, err
+	}
+
+	log.Printf("database: %s", pgOpts.Database)
```
Comment (stoksc) [inline master/internal/db/setup.go:101]: dont forget to remove this log since it would print credentials

File: master/internal/db/setup.go [hunk 2]
```diff
@@ -83,24 +85,64 @@ func Connect(opts *config.DBConfig) (*PgDB, error) {
 	return db, nil
 }
 
+// IsNew checks to see if the database's migration tracking tables have been
+// created, and if so, if it's above version 0. It returns `false` if the current
+// version exists and is higher than zero, `true` otherwise.
+// This is not guaranteed to be accurate if the database is otherwise in a bad or
+// incomplete state. If an error is returned, the bool should be ignored.
+func IsNew(opts *config.DBConfig) (bool, error) {
```
Comment (stoksc) [inline master/internal/db/setup.go:93]: we should add a test for this. i'd prefer a test that runs core.go::master.Run and tests the flow of creating a cluster many times without a password and the failure not going away. but testing this in isolation might make sense too.

File: master/internal/db/setup.go [hunk 3]
```diff
@@ -83,24 +85,64 @@ func Connect(opts *config.DBConfig) (*PgDB, error) {
...
+		return true, nil
+	}
+
+	collection := migrations.NewCollection()
+	collection.DisableSQLAutodiscover(true)
+	version, err := collection.Version(pgConn)
+	if err != nil {
+		return false, err
+	}
+	log.Printf("Database version is %d", version)
```
Comment (stoksc) [inline master/internal/db/setup.go:124]: i dont think we need this, db version is logged later. if it's useful for debugging maybe at debug level then.

**Review by stoksc — APPROVED** [review]
lgtm modulo feedback

File: master/internal/db/setup.go (hunk 2 above)
Comment (jesse-amano-hpe) [reply master/internal/db/setup.go:93]: Added a test; unfortunately running said test with a good-enough password allowed it to get all the way to running the service, which took me down a bit of a rabbit-hole figuring out how to shut the thing down.

Comment (stoksc) [reply master/internal/db/setup.go:93]: yeah, went i got to do the "removed actors from the agent" spike i had enough time to add graceful shutdown via context cancellation. master hasn't been so fortunate.

Comment (stoksc) [reply master/internal/db/setup.go:93]: oh jk, i think i (or someone) did some work towards this: https://github.com/determined-ai/determined/blame/main/master/internal/core.go#L1123. i'd be interested to learn more about what went south.

**Pull request merged**